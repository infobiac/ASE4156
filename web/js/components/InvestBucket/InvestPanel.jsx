// @flow
import React from 'react';
import Dialog, {
  DialogContent,
  DialogTitle,
  DialogActions,
} from 'material-ui/Dialog';
import Button from 'material-ui/Button';
import Table, { TableBody, TableCell, TableHead, TableRow } from 'material-ui/Table';
import Slider from 'rc-slider';

import StockGraph from '../StockGraph/StockGraph';

type Props = {
  available: number,
  bucket: {
    name: string,
    value: number,
    ownedAmount: number,
    history: Array<{date: Date, value: number}>
  },
  investFunc: number => void,
  cancelFunc: () => void,
}
type State = {
  investedAmount: number,
}

class InvestPanel extends React.Component<Props, State> {
  constructor() {
    super();
    this.state = {
      investedAmount: 0,
    };
  }
  render() {
    const values = [{
      name: this.props.bucket.name,
      data: this.props.bucket.history,
    }];
    const ownedQuantity = this.props.bucket.ownedAmount;
    const additionalQuantity = this.state.investedAmount / this.props.bucket.value;
    const ownedValue = this.props.bucket.ownedAmount * this.props.bucket.value;
    const additionalValue = this.state.investedAmount;
    const totalValue = ownedValue + additionalValue;

    return (
      <Dialog open>
        <DialogTitle>Invest into bucket</DialogTitle>
        <DialogContent>
          <StockGraph
            id="chart"
            title="Test Chart"
            quotes={values}
          />
          <Slider
            value={totalValue}
            id="invest-amount-slider"
            onChange={investedAmount => this.setState(() => ({
              investedAmount: investedAmount - ownedValue,
            }))}
            max={this.props.available + ownedValue}
            step={0.01}
          />
          <Table>
            <TableHead>
              <TableRow>
                <TableCell padding="dense">Name</TableCell>
                <TableCell padding="dense">Quantity</TableCell>
                <TableCell padding="dense">Value</TableCell>
                <TableCell padding="dense">Total Value</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              <TableRow>
                <TableCell padding="dense">{this.props.bucket.name} (Owned)</TableCell>
                <TableCell padding="dense">{ownedQuantity.toFixed(2)}</TableCell>
                <TableCell padding="dense">{this.props.bucket.value.toFixed(2)}</TableCell>
                <TableCell padding="dense">{ownedValue.toFixed(2)}</TableCell>
              </TableRow>
              <TableRow>
                <TableCell padding="dense">
                  {this.props.bucket.name}
                  {
                    additionalValue !== 0
                      ? <span>({additionalValue > 0 ? 'Buy' : 'Sell'})</span>
                      : null
                  }
                </TableCell>
                <TableCell padding="dense">{additionalQuantity.toFixed(2)}</TableCell>
                <TableCell padding="dense">{this.props.bucket.value.toFixed(2)}</TableCell>
                <TableCell padding="dense">{additionalValue.toFixed(2)}</TableCell>
              </TableRow>
              <TableRow>
                <TableCell padding="dense">Available</TableCell>
                <TableCell padding="dense" />
                <TableCell padding="dense" />
                <TableCell padding="dense">{(this.props.available - this.state.investedAmount).toFixed(2)}</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        </DialogContent>
        <DialogActions>
          <Button id="cancel" onClick={this.props.cancelFunc}>Cancel</Button>
          <Button id="save" onClick={() => this.props.investFunc(additionalQuantity)}>
            Invest
          </Button>
        </DialogActions>
      </Dialog>);
  }
}

export default InvestPanel;
