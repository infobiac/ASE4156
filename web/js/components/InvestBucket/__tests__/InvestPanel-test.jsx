import React from 'react';
import { shallow } from 'enzyme';
import InvestPanel from '../InvestPanel';

describe('InvestPanel', () => {
  it('exists', () => {
    expect(InvestPanel).toBeTruthy();
  });

  it('renders', () => {
    const comp = shallow(<InvestPanel
      available={100.0}
      bucket={{
        name: 'Some bucket', value: 3.0, ownedAmount: 2, history: [],
      }}
      investFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    expect(comp).toMatchSnapshot();
  });

  it('change amount (buy)', () => {
    const comp = shallow(<InvestPanel
      available={100.0}
      bucket={{
        name: 'Some bucket', value: 3.0, ownedAmount: 2, history: [],
      }}
      investFunc={jest.fn()}
      cancelFunc={jest.fn()}
    />);
    comp.find('#invest-amount-slider').props().onChange(15);
    comp.update();
    expect(comp).toMatchSnapshot();
  });

  it('cancel', () => {
    const cancel = jest.fn();
    const comp = shallow(<InvestPanel
      available={100.0}
      bucket={{
        name: 'Some bucket', value: 3.0, ownedAmount: 2, history: [],
      }}
      investFunc={jest.fn()}
      cancelFunc={cancel}
    />);
    expect(cancel.mock.calls.length).toEqual(0);
    comp.find('#cancel').simulate('click');
    expect(cancel.mock.calls.length).toEqual(1);
  });

  it('save (sell)', () => {
    const save = jest.fn();
    const comp = shallow(<InvestPanel
      available={100.0}
      bucket={{
        name: 'Some bucket', value: 3.0, ownedAmount: 2, history: [],
      }}
      investFunc={save}
      cancelFunc={jest.fn()}
    />);
    comp.find('#invest-amount-slider').props().onChange(0);
    comp.update();
    expect(comp).toMatchSnapshot();
    expect(save.mock.calls.length).toEqual(0);
    comp.find('#save').simulate('click');
    expect(save.mock.calls.length).toEqual(1);
    expect(save.mock.calls[0][0]).toEqual(-2);
  });
});

/*
type Props = {
  available: number,
  bucket: {
    name: string,
    value: number,
    ownedAmount: number,
    history: Array<{date: Date, value: number}>
  },
  investFunc: number => void,
  cancelFunc: () => void,
}
 */
