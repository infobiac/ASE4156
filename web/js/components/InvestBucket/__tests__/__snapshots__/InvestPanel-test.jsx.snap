// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestPanel change amount (buy) 1`] = `
<withStyles(Dialog)
  open={true}
>
  <withStyles(DialogTitle)>
    Invest into bucket
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <StockGraph
      compare="ABSOLUTE"
      height="400px"
      id="chart"
      quotes={
        Array [
          Object {
            "data": Array [],
            "name": "Some bucket",
          },
        ]
      }
      title="Test Chart"
    />
    <ComponentEnhancer(undefined)
      activeDotStyle={Object {}}
      className=""
      disabled={false}
      dotStyle={Object {}}
      dots={false}
      handle={[Function]}
      handleStyle={
        Array [
          Object {},
        ]
      }
      id="invest-amount-slider"
      included={true}
      marks={Object {}}
      max={106}
      min={0}
      onAfterChange={[Function]}
      onBeforeChange={[Function]}
      onChange={[Function]}
      prefixCls="rc-slider"
      railStyle={Object {}}
      step={0.01}
      trackStyle={
        Array [
          Object {},
        ]
      }
      value={15}
      vertical={false}
    />
    <withStyles(Table)>
      <withStyles(TableHead)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Name
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Quantity
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Value
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Total Value
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableHead)>
      <withStyles(TableBody)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Some bucket
             (Owned)
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            2.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            6.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Some bucket
            <span>
              (
              Buy
              )
            </span>
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            9.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Available
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          >
            91.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableBody)>
    </withStyles(Table)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      id="save"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`InvestPanel renders 1`] = `
<withStyles(Dialog)
  open={true}
>
  <withStyles(DialogTitle)>
    Invest into bucket
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <StockGraph
      compare="ABSOLUTE"
      height="400px"
      id="chart"
      quotes={
        Array [
          Object {
            "data": Array [],
            "name": "Some bucket",
          },
        ]
      }
      title="Test Chart"
    />
    <ComponentEnhancer(undefined)
      activeDotStyle={Object {}}
      className=""
      disabled={false}
      dotStyle={Object {}}
      dots={false}
      handle={[Function]}
      handleStyle={
        Array [
          Object {},
        ]
      }
      id="invest-amount-slider"
      included={true}
      marks={Object {}}
      max={106}
      min={0}
      onAfterChange={[Function]}
      onBeforeChange={[Function]}
      onChange={[Function]}
      prefixCls="rc-slider"
      railStyle={Object {}}
      step={0.01}
      trackStyle={
        Array [
          Object {},
        ]
      }
      value={6}
      vertical={false}
    />
    <withStyles(Table)>
      <withStyles(TableHead)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Name
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Quantity
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Value
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Total Value
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableHead)>
      <withStyles(TableBody)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Some bucket
             (Owned)
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            2.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            6.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Some bucket
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            0.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Available
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          >
            100.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableBody)>
    </withStyles(Table)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      id="save"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`InvestPanel save (sell) 1`] = `
<withStyles(Dialog)
  open={true}
>
  <withStyles(DialogTitle)>
    Invest into bucket
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <StockGraph
      compare="ABSOLUTE"
      height="400px"
      id="chart"
      quotes={
        Array [
          Object {
            "data": Array [],
            "name": "Some bucket",
          },
        ]
      }
      title="Test Chart"
    />
    <ComponentEnhancer(undefined)
      activeDotStyle={Object {}}
      className=""
      disabled={false}
      dotStyle={Object {}}
      dots={false}
      handle={[Function]}
      handleStyle={
        Array [
          Object {},
        ]
      }
      id="invest-amount-slider"
      included={true}
      marks={Object {}}
      max={106}
      min={0}
      onAfterChange={[Function]}
      onBeforeChange={[Function]}
      onChange={[Function]}
      prefixCls="rc-slider"
      railStyle={Object {}}
      step={0.01}
      trackStyle={
        Array [
          Object {},
        ]
      }
      value={0}
      vertical={false}
    />
    <withStyles(Table)>
      <withStyles(TableHead)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Name
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Quantity
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Value
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            Total Value
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableHead)>
      <withStyles(TableBody)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Some bucket
             (Owned)
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            2.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            6.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Some bucket
            <span>
              (
              Sell
              )
            </span>
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            -2.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            3.00
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          >
            -6.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
        <withStyles(TableRow)>
          <withStyles(TableCell)
            padding="dense"
          >
            Available
          </withStyles(TableCell)>
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          />
          <withStyles(TableCell)
            padding="dense"
          >
            106.00
          </withStyles(TableCell)>
        </withStyles(TableRow)>
      </withStyles(TableBody)>
    </withStyles(Table)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      id="save"
      onClick={[Function]}
    >
      Invest
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;
